pkg/client/reader_test.go

```diff
@@ -0,0 +1,230 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"sync"
+	"testing"
+	"time"
+
+	protobuf "github.com/golang/protobuf/proto"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+)
+
+func TestNewFromReader_Empty(t *testing.T) {
+	_, err := NewFromReader(bytes.NewBuffer([]byte("")), nil)
+	assert.Error(t, err)
+}
+
+func TestNewFromReader_BadProto(t *testing.T) {
+	_, err := NewFromReader(bytes.NewBuffer([]byte("bad proto")), nil)
+	assert.Error(t, err)
+}
+
+func TestNewFromReader_Connects(t *testing.T) {
+	ca, err := NewCA()
+	require.NoError(t, err)
+	peer, err := ca.GeneratePair()
+	require.NoError(t, err)
+	peerCert, err := peer.Certificate()
+	require.NoError(t, err)
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(ca.caPEM)
+	creds := credentials.NewTLS(&tls.Config{
+		Certificates: []tls.Certificate{peerCert},
+		ClientCAs:    caCertPool,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	})
+
+	var m sync.Mutex
+	token := "expected_token"
+	gotValid := false
+	srv := StubServer{
+		CheckinImpl: func(observed *proto.StateObserved) *proto.StateExpected {
+			m.Lock()
+			defer m.Unlock()
+
+			if observed.Token == token {
+				gotValid = true
+				return &proto.StateExpected{
+					State:          proto.StateExpected_RUNNING,
+					ConfigStateIdx: 1,
+					Config:         "config",
+				}
+			}
+			// disconnect
+			return nil
+		},
+		ActionImpl: func(response *proto.ActionResponse) error {
+			// actions not tested here
+			return nil
+		},
+		actions: make(chan *performAction, 100),
+	}
+	require.NoError(t, srv.Start(grpc.Creds(creds)))
+	defer srv.Stop()
+
+	connInfo := &proto.ConnInfo{
+		Addr:     fmt.Sprintf(":%d", srv.Port),
+		Token:    token,
+		CaCert:   ca.caPEM,
+		PeerCert: peer.Crt,
+		PeerKey:  peer.Key,
+	}
+	connInfoBytes, err := protobuf.Marshal(connInfo)
+	require.NoError(t, err)
+	connInfoReader := bytes.NewReader(connInfoBytes)
+
+	impl := &StubClientImpl{}
+	validClient, err := NewFromReader(connInfoReader, impl)
+	require.NoError(t, err)
+	require.NoError(t, validClient.Start(context.Background()))
+	defer validClient.Stop()
+	require.NoError(t, waitFor(func() error {
```

Comment:
Where is the `waitFor` function defined?

Reply:
Its defined in the client_test.go file that is in the same module